atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1991,104 +1979,102 @@ private void commitWrites(TransactionService transactionService) {
-                // in which the transaction key-value-service is invalid may also affect subsequent checks, and we
-                // would prefer for these to be flagged explicitly as such.
-                // Timed; this may in some implementations end up requiring external RPCs or database calls.
-                timedAndTraced(
-                        "transactionKvsValidityCheck",
-                        () -> throwIfTransactionKeyValueServiceNoLongerValid(commitTimestamp));
-
-                // Serializable transactions need to check their reads haven't changed, by reading again at
-                // commitTs + 1. This must happen before the lock check for thorough tables, because the lock check
-                // verifies the immutable timestamp hasn't moved forward - thorough sweep might sweep a conflict out
-                // from underneath us.
-                timedAndTraced(
-                        "readWriteConflictCheck", () -> throwIfReadWriteConflictForSerializable(commitTimestamp));
-
-                // Verify that our locks and pre-commit conditions are still valid before we actually commit;
-                // this throwIfPreCommitRequirementsNotMet is required by the transaction protocol for correctness.
-                // We check the pre-commit conditions first since they may operate similarly to read write conflict
-                // handling - we should check lock validity last to ensure that sweep hasn't affected the checks.
-                timedAndTraced(
-                        "userPreCommitCondition",
-                        () -> preCommitRequirementValidator.throwIfPreCommitConditionInvalidAtCommitOnWriteTransaction(
-                                writes, commitTimestamp));
-
-                // Not timed, because this just calls ConjureTimelockServiceBlocking.refreshLockLeases, and that is
-                // timed.
-                traced(
-                        "preCommitLockCheck",
-                        () -> preCommitRequirementValidator.throwIfImmutableTsOrCommitLocksExpired(commitLocksToken));
-
-                // Not timed, because this just calls TransactionService.putUnlessExists, and that is timed.
-                traced(
-                        "commitPutCommitTs",
-                        () -> putCommitTimestamp(commitTimestamp, commitLocksToken, transactionService));
-
-                long microsSinceCreation = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis() - timeCreated);
-                snapshotTransactionMetricFactory
-                        .getTimer("commitTotalTimeSinceTxCreation")
-                        .update(Duration.of(microsSinceCreation, ChronoUnit.MICROS));
-                snapshotTransactionMetricFactory
-                        .getHistogram(AtlasDbMetricNames.SNAPSHOT_TRANSACTION_BYTES_WRITTEN)
-                        .update(localWriteBuffer.getValuesByteCount());
-            } finally {
+            closer.register(() -> {
```

Comment:
I like this in general, we should move more things to just be owned by this guy, e.g. it always felt weird that releasing the immutable timestamp lock is done in a different place than closing the transaction. To me SnapshotTransaction BASICALLY should own every single resource and every single resource should be registered with it in one way or another.

Reply:
I refactored things to lean more on this.